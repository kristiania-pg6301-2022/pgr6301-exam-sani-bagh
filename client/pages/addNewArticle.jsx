import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormInput } from "../lib/formInput";
import { NewsApiContext } from "../newsApiContext";

export function AddNewArticle() {
  const { createArticle } = useContext(NewsApiContext);
  const [title, setTitle] = useState("");
  const [author, setAuthor] = useState("");
  const [topic, setTopic] = useState("");
  const [text, setText] = useState("");

  const navigate = useNavigate();

  function handleSubmit(event) {
    event.preventDefault();
    createArticle({ title, author, topic, text });
    navigate("/");
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Add new article</h1>

      <FormInput label={"Title:"} value={title} onChangeValue={setTitle} />
      <FormInput label={"Author:"} value={author} onChangeValue={setAuthor} />
      <FormInput label={"Topic:"} value={topic} onChangeValue={setTopic} />
      <FormInput label={"Text:"} value={text} onChangeValue={setText} />
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}
