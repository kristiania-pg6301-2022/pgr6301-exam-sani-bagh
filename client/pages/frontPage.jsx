import { Link } from "react-router-dom";
import React from "react";

export function FrontPage() {
  return (
    <div>
      <h1>News Database</h1>
      <ul>
        <li>
          <Link to={"/news"}>List news</Link>
        </li>
        <li>
          <Link to={"/login"}>Add new article</Link>
        </li>
      </ul>
    </div>
  );
}
