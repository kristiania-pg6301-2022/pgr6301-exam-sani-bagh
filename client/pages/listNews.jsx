import React, { useContext, useState } from "react";
import { useLoading } from "../useLoading";
import { NewsApiContext } from "../newsApiContext";
import { fetchJSON } from "../lib/fetchJSON";

function ArticleCard({ article: { title, author, category, text } }) {
  return (
    <>
      <h3>{title}</h3>
      <h3>Category: {category}</h3>
      <div>{author}</div>
      <p>{text}</p>
    </>
  );
}

export function ListNews() {
  const { listNews } = useContext(NewsApiContext);
  const [category, setCategory] = useState("");
  const [categoryQuery, setCategoryQuery] = useState("");
  const { loading, error, data } = useLoading(
    async () => await listNews({ category }),
    [category]
  );

  function handleSubmitQuery(e) {
    e.preventDefault();
    setCategory(categoryQuery);
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>News in the database</h1>

      <div>
        <form onSubmit={handleSubmitQuery}>
          <label>
            Topic:
            <input
              id="category-query"
              value={categoryQuery}
              onChange={(e) => setCategoryQuery(e.target.value)}
            />
            <button>Filter</button>
          </label>
        </form>
      </div>

      {data.map((article, index) => (
        <ArticleCard key={index} article={article} />
      ))}
    </div>
  );
}

export function ListAllNews() {
  const { listAllNews } = useContext(NewsApiContext);
  const [category, setCategory] = useState("");
  const { loading, error, data } = useLoading(
    async () => await listAllNews({ category }),
    [category]
  );

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>Daily posted Articles</h1>
      <ul>
        {data.map((article, index) => (
          <ArticleCard key={index} article={article} />
        ))}
      </ul>
    </div>
  );
}
