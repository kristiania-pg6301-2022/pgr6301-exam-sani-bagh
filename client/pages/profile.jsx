import { fetchJSON } from "../lib/fetchJSON";
import { useLoading } from "../useLoading";

export function Profile({ user }) {
  return (
    <div>
      <h1>Profile</h1>
      <pre>{JSON.stringify(user, undefined, "  ")}</pre>
      <h1>{user.name}</h1>
    </div>
  );
}
