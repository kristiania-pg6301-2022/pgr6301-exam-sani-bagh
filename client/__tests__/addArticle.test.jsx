import { MemoryRouter } from "react-router-dom";
import { AddNewArticle } from "../pages/addNewArticle";
import React from "react";
import ReactDOM from "react-dom";
import { NewsApiContext } from "../newsApiContext";
import { Simulate } from "react-dom/test-utils";

describe("add new article compoment", () => {
  it("should show article form", function () {
    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <AddNewArticle />
      </MemoryRouter>,
      domElement
    );
    expect(domElement.innerHTML).toMatchSnapshot();
    expect(
      Array.from(domElement.querySelectorAll("form label strong")).map(
        (e) => e.innerHTML
      )
    ).toEqual(["Title:", "Author:", "Topic:"]);
  });

  it("should add article on submit", function () {
    const createArticle = jest.fn();
    const title = "Test article";
    const domElement = document.createElement("div");
    ReactDOM.render(
      <NewsApiContext.Provider value={{ createArticle }}>
        <MemoryRouter>
          <AddNewArticle />
        </MemoryRouter>
      </NewsApiContext.Provider>,
      domElement
    );
    Simulate.change(domElement.querySelector(".form-input input"), {
      target: { value: title },
    });
    Simulate.change(
      domElement.querySelector(".form-input:nth-of-type(2) input"),
      {
        target: { value: "Test Author" },
      }
    );
    Simulate.submit(domElement.querySelector("form"));
    expect(createArticle).toBeCalledWith({
      title,
      author: "Test Author",
      topic: "",
    });
  });
});
