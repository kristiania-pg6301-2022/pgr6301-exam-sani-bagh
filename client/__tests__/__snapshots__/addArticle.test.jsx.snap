// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add new article compoment should show article form 1`] = `"<form><h1>Add new article</h1><div class=\\"form-input\\"><label><strong>Title:</strong> <input value=\\"\\"></label></div><div class=\\"form-input\\"><label><strong>Author:</strong> <input value=\\"\\"></label></div><div class=\\"form-input\\"><label><strong>Topic:</strong> <input value=\\"\\"></label></div><div class=\\"form-input\\"><label><strong>Text:</strong> <input value=\\"\\"></label></div><div><button>Submit</button></div></form>"`;
